강원대 BE 최현준 Step2-주문하기

Hyeonjun0527 opened the pull request:
### 특이한 점
OrderPersistenceAdapter의 save 메서드 : 연관 관계 저장에서 불필요한 쿼리를 하지 않기 위해 entityManager.getReference()를 사용.
OrderInteractor : 주문시에 카카오톡 메세지 발송 실패시 주문이 완료되지 않도록 하였음
  - 상품 옵션 및 회원 정보 검증
  - 주문 생성 및 저장
  - Option 재고 차감 로직 포함
  - 주문 완료 시 Wish에서 해당 항목 제거
  - 카카오 메시지 API로 주문 완료 메시지 발송

AuthResponse에 사용자 역할 필드 추가
역할 정보를 클라이언트에게 주지만, 실제 엔드포인트 호출 시 검사는 이루어지므로 역할정보를 클라이언트에서 저장해도 되겠다고 판단함.
- KakaoMessageApiClient 구현
- AppProperties 추가
  - 애플리케이션 기본 URL 설정 관리
- KakaoApiClientException 추가
- KakaoProperties에 messageMeUri 속성 추가

- Member 도메인에 kakaoAccessToken, kakaoRefreshToken 추가(아직 리프레시토큰 미도입 추후 step3에서 함께 도입을 목표로 하고 있음)


### 2차 리뷰

카카오 메시지 발송 로직 분리 및 주문 알림 비동기 처리
주문 확인 메시지 JSON 생성 로직 분리
(KakaoMessagePayloadFactory 추가)
- KakaoNotificationAdapter 추가
  - NotificationPort 구현을 통해 카카오톡 주문 알림 비동기처리.
- OrderInteractor NotificationPort 호출로 이름변경
- KakaoMessageApiClient에서 메시지 생성 기능을 PayloadFactory로 이동


kakaoInfo 도메인 도입.
- KakaoInfo 도메인 모델 도입
  - kakaoId, accessToken, refreshToken 관리
  - withTokens 메서드로 토큰 갱신 가능
- KakaoInfoEmbeddable JPA 임베디드 엔티티 추가
  - MemberEntity에 embedded로 변경하여 카카오 정보 저장
- Member 도메인 모델 Mapper, Repository 메서드 수정
  - 새로운 구조에 맞도록 수정

  추가

- 구현체 직접 의존한 것 리팩토링. KakaoAuthPort 인터페이스 추가
  - `fetchToken`, `fetchUserInfo` 메서드 정의
  - KakaoApiClient가 해당 인터페이스 구현
- MemberInteractor 가 인터페이스에 의존.


- NotificationEventListener 추가: 주문 완료 이벤트(OrderCompletedEvent) 처리하여 알림 비동기 전송
- OrderCompletedEvent 클래스 추가: 주문 ID와 회원 ID를 포함한 주문 완료 이벤트 정의
- OrderInteractor 수정:
  - 주문 완료 시 OrderCompletedEvent를 발행하도록 수정

주문 완료된 후에 알림을 비동기 전송하려면 이벤트 방식밖에 없는 것 같았습니다. 이것이 맞을까요?

Hyeonjun0527 committed:
refactor:스택트레이스 보이게 수정

Hyeonjun0527 committed:
feat(wish): 상품 옵션 관리 및 위시리스트 기능 확장

- Wish에 옵션 ID 및 옵션명 추가
- Product와 Option 간 관계 리팩토링
- Option 관리
  - OptionEntity에 updateOption 메서드 및 관계 설정 기능 추가
  - OptionEntity 및 Option 모델에 toString 메서드 추가
- ProductEntityMapper, WishEntityMapper 변경 사항 반영

Hyeonjun0527 committed:
feat(product, wish): 상품 및 위시 기능 확장 및 리팩토링

- 상품 옵션 관련 CreateOptionRequest 및 UpdateOptionRequest DTO 추가
- CreateProductRequest 및 UpdateProductRequest에 옵션 DTO 수정 적용
- WishRepository에 findByMemberIdAndOptionId 메서드 추가
- WishPersistenceAdapter에서 새로운 메서드 구현
- ProductPersistenceAdapter와 ProductInteractor에 EntityManager와 로그 추가
- 옵션 관련 로직 리팩토링:
  - ProductInteractor에서 옵션 ID를 직접 처리
  - updateProduct에서 기존 findOptionIdByName 메서드 제거
- 예외 타입 IllegalArgumentException -> NoSuchElementException으로 수정

Hyeonjun0527 committed:
feat(product, wish): 테스트 코드 작성

Hyeonjun0527 committed:
feat(product): 옵션 관리 기능 추가 및 상품 관련 로직 개선

- 상품 추가/수정 시 옵션(id, name, quantity) 관리 기능 구현
- 옵션 필드 추가 기능 및 삭제 기능
- 위시리스트 추가 시 옵션 정보 함꼐
- 관리자 상품 리스트에 옵션 정보 표시 및 드롭다운 구현
- 상품 추가 시 옵션 필드 초기화
- 관련 UI 스타일 및 레이아웃 업데이트

Hyeonjun0527 committed:
feat(member): 카카오 토큰 관리 필드 및 로직 추가

- Member 도메인에 kakaoAccessToken, kakaoRefreshToken 추가
- MemberEntity 및 MemberEntityMapper 변경
  - 카카오 엑세스 토큰과 리프레시 토큰 필드 추가 및 매핑 로직 적용
- loginWithKakao 유스케이스에서 토큰 갱신 로직 구현
  - 기존 멤버 객체에 토큰 정보를 반영 후 저장
- schema.sql에 Member 테이블의 토큰 필드 추가
- Order 테이블 정의 추가

Hyeonjun0527 committed:
feat(order): 주문 도메인 및 영속성 계층 추가

- 주문 도메인 모델 및 엔티티 추가
- OrderEntityMapper 추가로 도메인 모델과 엔티티 간 매핑 구현
- OrderJpaRepository 추가로 주문 엔티티 저장소 인터페이스 정의
- OrderPersistenceAdapter 구현으로 주문 데이터 저장소 기능 구현
- OrderRepository 포트 인터페이스 정의 및 Adapter 연결
- application.properties에 Kakao Message API 및 base-url 설정 추가
- Option 도메인에 decreaseQuantity 메서드 수정 및 이름과 수량을 업데이트하는 update 메서드 추가

Hyeonjun0527 committed:
feat(order): 카카오톡 주문 메시지 발송 기능 추가

- KakaoMessageApiClient 구현
- AppProperties 추가
  - 애플리케이션 기본 URL 설정 관리
- KakaoApiClientException 추가
- KakaoProperties에 messageMeUri 속성 추가
- OrderUseCase, PlaceOrderRequest, OrderResponse 구현
  - 옵션 ID로 옵션을 조회하는 기능 추가
- OptionPersistenceAdapter 수정
  - ProductEntity 대신 엔티티 참조를 사용해 Option 저장 로직 리팩토링

Hyeonjun0527 committed:
feat(order): 주문 도메인 추가 및 주문 기능 구현

- OrderController 작성
  - `POST /api/orders` 엔드포인트를 통해 주문 가능
  - @LoginMember로 인증된 회원 정보 전달받음
- OrderInteractor 작성
  - 상품 옵션 및 회원 정보 검증
  - 주문 생성 및 저장
  - Option 재고 차감 로직 포함
  - 주문 완료 시 Wish에서 해당 항목 제거
  - 카카오 메시지 API로 주문 완료 메시지 발송
- WishRepository/WishJpaRepository에 옵션 기반 삭제 메서드 추가
- schema.sql에서 테이블 및 외래키 정의 수정
  - Option 테이블 이름 수정 및 지연 무결성 관리 셋팅 추가
- application-dev.properties 설정 수정
  - spring.jpa.hibernate.ddl-auto를 none으로 변경
  - spring.devtools.restart.enabled 활성화

Hyeonjun0527 committed:
feat(order): 주문 메시지 추가 및 카카오톡 API 응답 처리 개선

- 위시리스트에 주문 메시지 작성 UI 추가 (HTML/CSS)
- 주문 기능 구현 및 메시지 입력값 전송 로직 추가 (JS, 엔드포인트 호출)
- KakaoApiClient에 오류 응답 핸들링 로직 추가
- AuthResponse에 사용자 역할 필드 추가
역할 정보를 클라이언트에게 주지만, 실제 엔드포인트 호출 시 검사는 이루어짐.
- OrderEntity 및 DB 스키마 컬럼 이름 소문자로 변경
- 기타 자잘한 UI 수정 및 로직 최적화

Hyeonjun0527 committed:
refactor(client): 카카오 API 클라이언트 구조 개선 및 코드 이전

- 기존 KakaoApiClient와 KakaoMessageApiClient를 infra/client/kakao 패키지로 이동
- RestClient 설정 분리 및 AppConfig에 RestClient 관련 Bean 추가
  - 기본 RestClient와 카카오 전용 RestClient 분리

Hyeonjun0527 committed:
fix(config/product): RestClient timeout 수정 및 상품 저장 로직 변경

- ClientConfig의 kakaoRestClient connectTimeout을 3초에서 7초로 변경
- ProductPersistenceAdapter에서 상품 및 옵션 저장 시 EntityManager의 merge 대신 JpaRepository의 save로 변경
  - 상품 저장 로직 - ProductJpaRepository.save() 사용
  - 옵션 저장 로직 - OptionJpaRepository.save() 사용

jjongwa commented on src/main/java/gift/order/application/usecase/OrderInteractor.java:
```diff
@@ -0,0 +1,76 @@
+        );
+        Order savedOrder = orderRepository.save(order);
+
+        // 4. 위시리스트에서 해당 상품 삭제
+        wishRepository.deleteByMemberIdAndOptionId(memberId, request.optionId());
+
+        // 5. 카카오톡 메시지 발송
+        if (member.kakaoAccessToken() != null) {
+            kakaoMessagePort.sendOrderConfirmationMessage(member.kakaoAccessToken(), savedOrder, product, option);
+        }
```
현재 재고 차감, 주문 저장, 위시리스트 상품 삭제, 메시지 전송이 모두 하나의 트랜잭션에 묶여 있는데요. 이러한 상황이 충분히 합리적일지 고민해 보면 좋을 것 같습니다.!

만약 카카오 메시지 호출하는 과정의 네트워크에 이상이 생겨서 전송이 특정 기간동안 모두 실패하는 상황이라면, 사용자 입장에선 전혀 주문을 할 수 없고, 서비스 입장에선 해당 기간동안 핵심 기능이 마비되는 상황일텐데(메시지 때문에 특정 기간동안 매출이 0이겠군요..!) 이런 방향성이 모두에게 올바를지, 어떻게 하면 이러한 싱황을 해결할 수 있을지 고민해 보면 좋을 것 같네요.!

Hyeonjun0527 replied:
아아 넵 감사합니다. 그런 생각은 못해봤네요. 반드시 메세지가 호출되어야한다고만 생각을 했는데, 핵심 기능 마비가 될 수가 있는 줄은 생각을 못했습니다!

jjongwa commented on src/main/java/gift/common/config/ClientConfig.java:
```diff
@@ -0,0 +1,29 @@
+public class ClientConfig {
+
+    @Bean
+    @Primary
```
`@Primary` 설정을 사용해 주셨네요.!
해당 어노테이션은 어떤 효과를 가지고, 현재 상황에선 어떤 기능을 하고 있나요?

Hyeonjun0527 replied:
RestClient를 생성자 주입할때 디폴트로 주입되도록 Primary를 설정하였습니다. 만약에 Primary를 붙이지 않으면 RestClient의 후보로 등록되는 빈이 2개가 되어버려서 에러가 나기 때문입니다!

jjongwa commented on src/test/java/gift/product/adapter/persistence/entity/ProductOptionUpdateTest.java:
```diff
@@ -0,0 +1,99 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@SpringBootTest
+@Transactional
+@TestPropertySource(properties = "spring.jpa.show-sql=true")
+public class ProductOptionUpdateTest {
```
`@SpringBootTest` 말고 다른 어노테이션으론 본 테스트가 동작할까요?

조금 더 작은 단위에서의 테스트의 필요성은 없을까요?

Hyeonjun0527 replied:
현재 레포지토리만 테스트하고 있어, @DataJpaTest로 바꿔 테스트해도 괜찮을 것 같네요 감사합니다.

jjongwa commented on src/main/java/gift/order/adapter/persistence/adapter/OrderPersistenceAdapter.java:
```diff
@@ -0,0 +1,33 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class OrderPersistenceAdapter implements OrderRepository {
+
+    private final OrderJpaRepository orderJpaRepository;
+    private final EntityManager entityManager;
+
+    public OrderPersistenceAdapter(OrderJpaRepository orderJpaRepository, EntityManager entityManager) {
+        this.orderJpaRepository = orderJpaRepository;
+        this.entityManager = entityManager;
+    }
+
+    @Override
+    public Order save(Order order) {
+        MemberEntity memberEntityProxy = entityManager.getReference(MemberEntity.class, order.memberId());
+        OptionEntity optionEntityProxy = entityManager.getReference(OptionEntity.class, order.optionId());
+
+        OrderEntity orderEntity = OrderEntityMapper.toEntity(order, memberEntityProxy, optionEntityProxy);
+        OrderEntity savedEntity = orderJpaRepository.save(orderEntity);
```
entityManager를 추가해 주셨는데, getReference 사용을 위한 거겠죠.!

현재 해당 부분의 코드가 성능 최적화가 꼭 필요한 병목 구간일까요?

기존 JPA 제공 로직 대신 entityManager를 통한 기능을 사용했을 때의 트레이드오프는 어떤게 있을까요?

Hyeonjun0527 replied:
getReference를 사용하게 된 계기는 서비스쪽(OrderInteractor)에서 이미 가져온 member와 optionId를 또다시 find하기가 부담스럽고 이 member와 option을 find하는 것을 완전히 어댑터로 옮기기도 부담스러운 상황이였어서 채택하게 되었습니다. 필요없는 SELECT 쿼리를 또 날리는 것이 리팩토링 대상이 될 것이라고 생각했기 때문입니다. 그래서 getReference로 id 만 가지고 나머지 필드는 비어있는 영속상태 객체를 만들었습니다. 그래서 Order 객체를 save 할때 id만 필요하므로 추가적인 쿼리를 생성하지 않을 수 있는 효과를 얻었습니다

entityManager를 사용할 시에는 엔티티의 생명주기를 수동적으로 다뤄야하는 트레이드 오프가 있을 것 같습니다.
getReference는 딱히 상황을 복잡하게 만들지 않지만, 
기존 JPA의 save 대신 만약 entityManager가 제공하는 merge나 persist를 사용한다면 더욱 조심해야 될 것이라고 생각합니다. save는 id가 존재하는지 아닌지 기준으로 merge하거나 persist를 하는데 이를 모르고 단순히 persist만 하는 식으로 save메서드를 구성하면 문제가 생길 것 같습니다

jjongwa commented on src/main/java/gift/member/domain/model/Member.java:
```diff
@@ -2,23 +2,36 @@
+        Role role,
+        LocalDateTime createdAt,
+        Long kakaoId,
+        String kakaoAccessToken,
+        String kakaoRefreshToken
```
요정도는 하나로 묶어서 사용하는게 더 응집성을 높일 수 있겠네요 (필요하다면 kakaoId까지 포함해서)

Hyeonjun0527 replied:
감사합니다

jjongwa commented on src/main/java/gift/infra/client/kakao/KakaoMessageApiClient.java:
```diff
@@ -0,0 +1,121 @@
+            .body(body)
+            .retrieve()
+            .onStatus(statusCode -> statusCode.isError(), (request, responseEntity) -> {
+                String responseBody;
+                try {
+                    responseBody = new String(responseEntity.getBody().readAllBytes());
+                } catch (IOException e) {
+                    throw new KakaoApiClientException("카카오 API 응답을 읽는 중 오류가 발생했습니다.", e);
+                }
+                log.error("카카오 메시지 전송 실패. status: {}, body: {}", responseEntity.getStatusCode(),
+                    responseBody);
+                throw new KakaoApiClientException("카카오 메시지 전송에 실패했습니다. 응답: " + responseBody);
+            })
+            .toEntity(String.class);
```
음 메서드 분리를 통해 조금 더 깔끔하게 구성할 수 있어 보여요.!
String responseBody; 를 미리 선언하고 이후에 대입하는 것이 아닌 상황에 따라 해당하는 responseBody를 만들어 바로 반환하는 편이 조금 더 명확하겠네요

Hyeonjun0527 replied:
넵 메서드 분리해보도록 하겠습니다!

jjongwa commented on src/main/java/gift/infra/client/kakao/KakaoMessageApiClient.java:
```diff
@@ -0,0 +1,121 @@
+        }
+    }
+
+    private String createMessageJson(Order order, Product product, Option option) {
```
요 부분도 필요하다면 따로 클래스를 분리해 사용할 수도 있겠네요.!

어떻게 생각하느냐에 따라 다르겠지만 카카오 api 통신을 하는 것과 통신을 위한 json param을 만드는 건 다른 책임으로 볼 수도 있으니까요.!

Hyeonjun0527 replied:
이부분도 반영해볼게요

jjongwa commented on src/main/java/gift/order/adapter/persistence/mapper/OrderEntityMapper.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    public static OrderEntity toEntity(Order domain, MemberEntity memberEntity, OptionEntity optionEntity) {
+        if (domain == null) {
+            return null;
+        }
+        return new OrderEntity(
+                domain.id(),
+                memberEntity,
+                optionEntity,
+                domain.productName(),
+                domain.optionName(),
+                domain.price(),
+                domain.quantity(),
+                domain.message(),
+                domain.orderDateTime()
+        );
```
domain에 대해서만 null check를 하는 이유가 있을까요?
OrderEntity의 member와 option엔 nullable=false로 되어 있던데, 해당 두 객체틑 null check가 필요할까요?

+mapper 단에서 체크가 필요할까요?

Hyeonjun0527 replied:
이론상으로는 레포지토리(데이터접근계층)은 서비스계층을 전혀 몰라야 하기 때문에 모든 계층에서 null체크를 해야한다고 생각하고 도메인또한 비지니스 규칙을 담당하기에 중복적으로 검증하는 것은 좋은 방식이라고 생각하고 있습니다! 엔티티에 있는 nullable은 데이터베이스에 어떤 데이터를 허용할 지 아닐지에 대한 것이니 안정성을 위해서 또한 필요할 거 같습니다.

퍼시스턴스 계층에서 처음 온 데이터를 퍼시스턴스 매퍼가 변환하니 null을 검사하는 것이 좋다고 생각했습니다!

jjongwa reviewed (changes requested):
2단계 잘 구현해 주셨어요!
기간 안에 제출하는 것도 중요한 사항이니 완벽한 코드 보단 기본 요구사항을 만족하고 정상 작동되기만 한다면 바로바로 리뷰 요청 하는걸 추천드립니다..!

Hyeonjun0527 committed:
refactor(member): kakaoInfo 도메인 도입.
- KakaoInfo 도메인 모델 도입
  - kakaoId, accessToken, refreshToken 관리
  - withTokens 메서드로 토큰 갱신 가능
- KakaoInfoEmbeddable JPA 임베디드 엔티티 추가
  - MemberEntity에 embedded로 변경하여 카카오 정보 저장
- Member 도메인 모델 Mapper, Repository 메서드 수정
  - 새로운 구조에 맞도록 수정

  추가

- 구현체 직접 의존한 것 리팩토링. KakaoAuthPort 인터페이스 추가
  - `fetchToken`, `fetchUserInfo` 메서드 정의
  - KakaoApiClient가 해당 인터페이스 구현
- MemberInteractor 가 인터페이스에 의존.

Hyeonjun0527 committed:
feat(notification): 카카오 메시지 발송 로직 분리 및 주문 알림 비동기 처리
주문 확인 메시지 JSON 생성 로직 분리
(KakaoMessagePayloadFactory 추가)

- KakaoNotificationAdapter 추가
  - NotificationPort 구현을 통해 카카오톡 주문 알림 비동기처리.

- OrderInteractor NotificationPort 호출로 이름변경

- KakaoMessageApiClient에서 메시지 생성 기능을 PayloadFactory로 이동

Hyeonjun0527 committed:
test(persistence): jpa 연관관계 학습용 테스트 케이스 추가.

Hyeonjun0527 committed:
refactor(database): 토큰 컬럼 이름 변경 및 초기 데이터 수정

- kakao_access_token, kakao_refresh_token 컬럼명을 access_token, refresh_token으로 변경

Hyeonjun0527 committed:
feat(notification): 주문 완료 시 알림 전송을 위한 이벤트 리스너 추가

- NotificationEventListener 추가: 주문 완료 이벤트(OrderCompletedEvent) 처리하여 알림 비동기 전송
- OrderCompletedEvent 클래스 추가: 주문 ID와 회원 ID를 포함한 주문 완료 이벤트 정의
- OrderInteractor 수정:
  - NotificationPort 의존성 삭제
  - 주문 완료 시 OrderCompletedEvent를 발행하도록 수정
  - ApplicationEventPublisher 의존성 주입 및 활용

jjongwa reviewed (approved):
리뷰 반영 잘 해주셨어요!
몇몇 질문엔 답변이 없는데, 다시 한번 확인한 뒤에 넘어가 주세요~!
